README
Інформаційна система "Стрічка" - дебагер протоколу

Версія: a2

ПРО СИСТЕМУ "СТРІЧКА"

Система "Стрічка" - це новітня система обробки повідомлень для 
інформаційних агентств та інших видань.

Докладніше про систему "Стрічка" можно повідати зі 
специфікацій у репозиторії RibbonSpecs.

https://github.com/spoilt-exile/RibbonSpecs

ПРО ЦЕЙ РЕПОЗИТОРІЙ

Цей репозиторій містить код дебагеру протоколу системи
"Стрічка". Дебагер протоколу необхіден для з’єднання з сервером
і перевірки його функціональних можливостей. Дебагер підтримує 
базовий протокол, протокол керування сервером поки що
розробляється.

АВТОРСЬКІ ПРАВА ТА УМОВИ РОЗПОВСЮДЖЕННЯ

Код системи розповсюджується за ліцензією GNU GPLv2 або за 
умовами більш нової ліцензії GNU GPL.

© Непочатов Станіслав 2012-2013.

CHANGELOG
===============================================================
RibbonProber (Дебагер протоколу "Стрічка") a1 (24.06.2012)
===============================================================
 + реалізована підтримка базового протоколу;
 + реалізована експериментальна підтримка протоколу керування
   сервером;
 + додана довідка до кожної команди протоколу;
 + додана автовставка аргументів команд;
===============================================================
RibbonProber (Дебагер протоколу "Стрічка") a2 (27.10.2013)
 + додано виявленя кодової сторінки для команди 
   RIBBON_NCTL_INIT:;
 + Додання команд для роботи з системними ознаками;
 + Додання команд для видаленого режиму;
 * Оновлення у опису команди RIBBON_POST_MESSAGE:;
 * Оновлення у опису команди RIBBON_MODIFY_MESSAGE:;
 * Перебудова протоколу рівня 0;
 * Змінено вставку для команди логіну;
 - видалено підтримку команди RIBBON_SCTL_LOGIN:;
 Х Виправлено помилки у опису команд;
===============================================================

src/RibbonProber/mainFrame.java
/**
 * This file is part of RibbonProber application (check README).
 * Copyright (C) 2012-2013 Stanislav Nepochatov
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
**/

package RibbonProber;

/**
 * Main frame class
 * @author Stanislav Nepochatov
 */
public class mainFrame extends javax.swing.JFrame {
    
    /** Creates new form mainFrame */
    public mainFrame() {
        initComponents();
    }
    
    /**
     * Add message to gui list
     * @param message 
     */
    public synchronized void addToLog(String message) {
        this.logArea.setText(
                this.logArea.getText()  + message);
        this.logArea.setCaretPosition(this.logArea.getText().length());
    }

    /** This method is called from within the constructor to
     * initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is
     * always regenerated by the Form Editor.
     */
    @SuppressWarnings("unchecked")
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        commandBox = new javax.swing.JComboBox();
        jScrollPane1 = new javax.swing.JScrollPane();
        commandField = new javax.swing.JTextPane();
        launchBut = new javax.swing.JButton();
        ipField = new javax.swing.JTextField();
        connectBut = new javax.swing.JButton();
        portField = new javax.swing.JTextField();
        jLabel1 = new javax.swing.JLabel();
        jLabel2 = new javax.swing.JLabel();
        jLabel3 = new javax.swing.JLabel();
        pasteBut = new javax.swing.JButton();
        flushBut = new javax.swing.JButton();
        jScrollPane2 = new javax.swing.JScrollPane();
        logArea = new javax.swing.JTextArea();
        helpBut = new javax.swing.JButton();
        flushJournalBut = new javax.swing.JButton();
        levelBox = new javax.swing.JComboBox();

        setDefaultCloseOperation(javax.swing.WindowConstants.EXIT_ON_CLOSE);
        setTitle("Дебагер протоколу системи \"Стрічка\" a2");

        commandBox.setModel(new javax.swing.DefaultComboBoxModel(ProtocolDecription.getCommandList(-1)));
        commandBox.setToolTipText("");

        jScrollPane1.setViewportView(commandField);

        launchBut.setForeground(new java.awt.Color(153, 0, 0));
        launchBut.setText("Послати!");
        launchBut.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                launchButActionPerformed(evt);
            }
        });

        ipField.setText("127.0.0.1");

        connectBut.setForeground(new java.awt.Color(204, 102, 0));
        connectBut.setText("З'єднати");
        connectBut.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                connectButActionPerformed(evt);
            }
        });

        portField.setText("3003");

        jLabel1.setText("Команда");

        jLabel2.setText("Адреса сервера");

        jLabel3.setText("Порт");

        pasteBut.setText("Вставити");
        pasteBut.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                pasteButActionPerformed(evt);
            }
        });

        flushBut.setText("Відчистити");
        flushBut.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                flushButActionPerformed(evt);
            }
        });

        logArea.setColumns(20);
        logArea.setEditable(false);
        logArea.setRows(5);
        logArea.setCursor(new java.awt.Cursor(java.awt.Cursor.TEXT_CURSOR));
        jScrollPane2.setViewportView(logArea);

        helpBut.setText("?");
        helpBut.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                helpButActionPerformed(evt);
            }
        });

        flushJournalBut.setText("Відчистити журнал");
        flushJournalBut.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                flushJournalButActionPerformed(evt);
            }
        });

        levelBox.setModel(new javax.swing.DefaultComboBoxModel(new String[] { "Всі команди протоколу", "Команди з’єднання [LEVEL_0]", "Базові команди [LEVEL_1]", "Команди до сервера [LEVEL_2]" }));
        levelBox.addItemListener(new java.awt.event.ItemListener() {
            public void itemStateChanged(java.awt.event.ItemEvent evt) {
                levelBoxItemStateChanged(evt);
            }
        });

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(getContentPane());
        getContentPane().setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(jScrollPane2)
                    .addGroup(layout.createSequentialGroup()
                        .addComponent(jLabel2)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addComponent(ipField, javax.swing.GroupLayout.PREFERRED_SIZE, 106, javax.swing.GroupLayout.PREFERRED_SIZE)
                        .addGap(18, 18, 18)
                        .addComponent(jLabel3)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addComponent(portField)
                        .addGap(18, 18, 18)
                        .addComponent(flushJournalBut)
                        .addGap(18, 18, 18)
                        .addComponent(connectBut))
                    .addGroup(javax.swing.GroupLayout.Alignment.TRAILING, layout.createSequentialGroup()
                        .addComponent(jScrollPane1, javax.swing.GroupLayout.DEFAULT_SIZE, 675, Short.MAX_VALUE)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING, false)
                            .addComponent(flushBut, javax.swing.GroupLayout.DEFAULT_SIZE, 126, Short.MAX_VALUE)
                            .addComponent(launchBut, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                            .addComponent(pasteBut, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)))
                    .addGroup(javax.swing.GroupLayout.Alignment.TRAILING, layout.createSequentialGroup()
                        .addComponent(levelBox, 0, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                        .addComponent(jLabel1)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                        .addComponent(commandBox, javax.swing.GroupLayout.PREFERRED_SIZE, 406, javax.swing.GroupLayout.PREFERRED_SIZE)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addComponent(helpBut)))
                .addContainerGap())
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(ipField, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(connectBut)
                    .addComponent(portField, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(jLabel2)
                    .addComponent(jLabel3)
                    .addComponent(flushJournalBut))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                .addComponent(jScrollPane2, javax.swing.GroupLayout.DEFAULT_SIZE, 440, Short.MAX_VALUE)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(commandBox, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(jLabel1)
                    .addComponent(helpBut)
                    .addComponent(levelBox, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addGroup(layout.createSequentialGroup()
                        .addComponent(launchBut)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addComponent(pasteBut)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addComponent(flushBut))
                    .addComponent(jScrollPane1, javax.swing.GroupLayout.PREFERRED_SIZE, 101, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addContainerGap())
        );

        pack();
    }// </editor-fold>//GEN-END:initComponents

    private void connectButActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_connectButActionPerformed
        RibbonProber.connect();
    }//GEN-LAST:event_connectButActionPerformed

    private void launchButActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_launchButActionPerformed
        if (RibbonProber.outStream != null) {
            this.addToLog("> " + this.commandField.getText() + "\n");
            RibbonProber.outStream.println(this.commandField.getText());
        } else {
            this.addToLog("ПОМИЛКА: немає з'єднання з сервером!\n");
        }
    }//GEN-LAST:event_launchButActionPerformed

    private void pasteButActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_pasteButActionPerformed
        if (this.commandField.getText().isEmpty()) {
            this.commandField.setText(this.commandBox.getSelectedItem() +
                    ProtocolDecription.getPaste((String) this.commandBox.getSelectedItem()));
        } else {
            this.commandField.setText(this.commandField.getText() + "\n"
                + this.commandBox.getSelectedItem() +
                    ProtocolDecription.getPaste((String) this.commandBox.getSelectedItem()));
        }
    }//GEN-LAST:event_pasteButActionPerformed

    private void flushButActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_flushButActionPerformed
        this.commandField.setText("");
    }//GEN-LAST:event_flushButActionPerformed

    private void helpButActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_helpButActionPerformed
        String formatLine = "=====================================================";
        this.addToLog(formatLine + "\nДОВІДКА\n" + formatLine + "\n" + ProtocolDecription.getDoc((String) this.commandBox.getSelectedItem()) + "\n" + formatLine + "\n");
    }//GEN-LAST:event_helpButActionPerformed

    private void flushJournalButActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_flushJournalButActionPerformed
        this.logArea.setText("");
    }//GEN-LAST:event_flushJournalButActionPerformed

    private void levelBoxItemStateChanged(java.awt.event.ItemEvent evt) {//GEN-FIRST:event_levelBoxItemStateChanged
        this.commandBox.setModel(new javax.swing.DefaultComboBoxModel(ProtocolDecription.getCommandList(this.levelBox.getSelectedIndex() - 1)));
    }//GEN-LAST:event_levelBoxItemStateChanged

    /**
     * @param args the command line arguments
     */
    public static void main(String args[]) {
        /* Set the Nimbus look and feel */
        //<editor-fold defaultstate="collapsed" desc=" Look and feel setting code (optional) ">
        /* If Nimbus (introduced in Java SE 6) is not available, stay with the default look and feel.
         * For details see http://download.oracle.com/javase/tutorial/uiswing/lookandfeel/plaf.html 
         */
        try {
            for (javax.swing.UIManager.LookAndFeelInfo info : javax.swing.UIManager.getInstalledLookAndFeels()) {
                if ("Nimbus".equals(info.getName())) {
                    javax.swing.UIManager.setLookAndFeel(info.getClassName());
                    break;
                }
            }
        } catch (ClassNotFoundException ex) {
            java.util.logging.Logger.getLogger(mainFrame.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
        } catch (InstantiationException ex) {
            java.util.logging.Logger.getLogger(mainFrame.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
        } catch (IllegalAccessException ex) {
            java.util.logging.Logger.getLogger(mainFrame.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
        } catch (javax.swing.UnsupportedLookAndFeelException ex) {
            java.util.logging.Logger.getLogger(mainFrame.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
        }
        //</editor-fold>

        /* Create and display the form */
        java.awt.EventQueue.invokeLater(new Runnable() {

            public void run() {
                new mainFrame().setVisible(true);
            }
        });
    }
    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JComboBox commandBox;
    private javax.swing.JTextPane commandField;
    private javax.swing.JButton connectBut;
    private javax.swing.JButton flushBut;
    private javax.swing.JButton flushJournalBut;
    private javax.swing.JButton helpBut;
    public javax.swing.JTextField ipField;
    private javax.swing.JLabel jLabel1;
    private javax.swing.JLabel jLabel2;
    private javax.swing.JLabel jLabel3;
    private javax.swing.JScrollPane jScrollPane1;
    private javax.swing.JScrollPane jScrollPane2;
    private javax.swing.JButton launchBut;
    private javax.swing.JComboBox levelBox;
    private javax.swing.JTextArea logArea;
    private javax.swing.JButton pasteBut;
    public javax.swing.JTextField portField;
    // End of variables declaration//GEN-END:variables
}


src/RibbonProber/ProtocolDecription.java
/**
 * This file is part of RibbonProber application (check README).
 * Copyright (C) 2012-2013 Stanislav Nepochatov
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
**/

package RibbonProber;

/**
 * Class of protocol commands description and common tools
 * @author Stanislav Nepochatov
 */
public abstract class ProtocolDecription {
    
    /**
     * Commands array
     */
    private static java.util.ArrayList<ProtocolCommand> commands = new java.util.ArrayList<>();
    
    /**
     * Command class
     */
    private static class ProtocolCommand {
        
        /**
         * Name of command;
         */
        public String NAME;
        
        /**
         * Doc string of command;
         */
        public String DOC;
        
        /**
         * Auto paste string
         */
        public String AUTO_PASTE;
        
        /**
         * Protocol level of command
         */
        public Integer COMMAND_LEVEL;
        
        /**
         * Constructor for command
         * @param givenName name of command
         * @param givenDoc doc for command
         * @param givenPaste paste arguments for command
         */
        ProtocolCommand(String givenName, String givenDoc, String givenPaste, Integer givenLevel) {
            NAME = givenName;
            DOC = givenDoc;
            AUTO_PASTE = givenPaste;
            COMMAND_LEVEL = givenLevel;
        }
    }
    
    /**
     * Init protocol commands
     */
    public static void initProtocol() {
        
        commands.add(new ProtocolCommand("RIBBON_NCTL_INIT",
                "[LEVEL_0] RIBBON_NCTL_INIT\n"
                    + "Команда привітання з сервером.\n"
                    + "Аргументи: ТИП_З’ЄДНАННЯ,ВЕРСІЯ_ПРОТОКОЛУ,КОДОВА_СТОРІНКА\n"
                    + "Приклад: RIBBON_NCTL_INIT:CLIENT,a2,UTF-8",
                ":CLIENT,a2," + System.getProperty("file.encoding") , 0));
        
        commands.add(new ProtocolCommand("RIBBON_NCTL_LOGIN",
                "[LEVEL_0] RIBBON_NCTL_LOGIN\n" 
                    + "Команда реєстрації користувача у системі.\n"
                    + "Аргументи: ЛОГІН,ПАРОЛЬ у формі csv\n"
                    + "Приклад: RIBBON_NCTL_LOGIN:{root},74cc1c60799e0a786ac7094b532f01b1",
                ":{root},63a9f0ea7bb98050796b649e85481845", 0));
        
        commands.add(new ProtocolCommand("RIBBON_NCTL_GET_ID",
                "[LEVEL_0] RIBBON_NCTL_GET_ID\n"
                    + "Команда отримання ідентефикатору цієї сесії.\n"
                    + "Аргументи: немає",
                ":", 0));
        
        commands.add(new ProtocolCommand("RIBBON_NCTL_RESUME",
                "[LEVEL_0] RIBBON_NCTL_GET_ID\n"
                    + "Команда отримання ідентефикатору цієї сесії.\n"
                    + "Аргументи: ІДЕНТЕФІКАТОР\n"
                    + "Приклад: RIBBON_NCTL_RESUME:74cc1c60799e0a786ac7094b532f01b1",
                ":", 0));
        
        commands.add(new ProtocolCommand("RIBBON_NCTL_REM_LOGIN",
                "[LEVEL_0] RIBBON_NCTL_REM_LOGIN\n" 
                    + "Команда реєстрації користувача через видалений інтерфейс.\n"
                    + "Аргументи: {ЛОГІН},ПАРОЛЬ (у формі csv)\n"
                    + "Приклад: RIBBON_NCTL_REM_LOGIN:{root},63a9f0ea7bb98050796b649e85481845",
                ":{root},63a9f0ea7bb98050796b649e85481845", 0));
        
        commands.add(new ProtocolCommand("RIBBON_NCTL_GET_USERNAME",
                "[LEVEL_0] RIBBON_NCTL_GET_USERNAME\n"
                    + "Команда отримання логіну поточного користувача.\n"
                    + "Аргументи: немає",
                ":", 0));
        
        commands.add(new ProtocolCommand("RIBBON_NCTL_SET_REMOTE_MODE",
                "[LEVEL_0] RIBBON_NCTL_SET_REMOTE_MODE\n"
                    + "Команда керування видаленим режимом поточного з’єднання.\n"
                    + "Аргументи: 0 чи 1 (0 - вимикає, 1- вмикає)\n"
                    + "Приклад: RIBBON_NCTL_SET_REMOTE_MODE:1",
                ":1", 0));
        
        commands.add(new ProtocolCommand("RIBBON_NCTL_ACCESS_CONTEXT",
                "[LEVEL_0] RIBBON_NCTL_ACCESS_CONTEXT\n"
                    + "Команда зміни контексту виконання наступної команди.\n"
                    + "Аргументи: {КОРИСТУВАЧ}\n"
                    + "Приклад: RIBBON_NCTL_ACCESS_CONTEXT:{test}"
                    + "\n\nУВАГА! Ця команда для виконання вимагає увікнення видаленого режиму\n"
                    + "(команда RIBBON_NCTL_SET_REMOTE_MODE)!",
                ":{test}", 0));
        
        commands.add(new ProtocolCommand("RIBBON_NCTL_CLOSE",
                "[LEVEL_0] RIBBON_NCTL_CLOSE\n"
                    + "Команда закриття з’єднання з сервером.\n"
                    + "Аргументи: немає",
                ":", 0));
        
        commands.add(new ProtocolCommand("RIBBON_GET_DIRS",
                "[LEVEL_1] RIBBON_GET_DIRS\n"
                    + "Команда повертання списку напрямків системи у csv формі.\n"
                    + "Аргументы: немає",
                ":", 1));
        
        commands.add(new ProtocolCommand("RIBBON_GET_PSEUDO",
                "[LEVEL_1] RIBBON_GET_PSEUDO\n"
                    + "Команда повернення списку псевдонапрямків які може використовувати\n"
                    + "користувач у сsv формі.\n"
                    + "Аргументи: немає"
                    + "\n\nУВАГА! Ця команда для виконання вимагає увікнення видаленого режиму\n"
                    + "(команда RIBBON_NCTL_SET_REMOTE_MODE)!",
                ":", 1));
        
        commands.add(new ProtocolCommand("RIBBON_GET_TAGS",
                "[LEVEL_1] RIBBON_GET_TAGS\n"
                    + "Команда повертення списку тегів системи у csv формі.\n"
                    + "Аргументи: немає",
                ":", 1));
        
        commands.add(new ProtocolCommand("RIBBON_LOAD_BASE_FROM_INDEX",
                "[LEVEL_1] RIBBON_LOAD_BASE_FROM_INDEX\n"
                    + "Команда повертання списку індекса бази повідомлень.\n"
                    + "починаючи з певного індексу.\n"
                    + "Аргументи:ИНДЕКС або 0 (повернення усього індексу бази)\n"
                    + "Приклад: RIBBON_LOAD_BASE_FROM_INDEX:0000000107",
                ":0", 1));
        
        commands.add(new ProtocolCommand("RIBBON_POST_MESSAGE",
                "[LEVEL_1] RIBBON_POST_MESSAGE\n"
                    + "Команда випуску повідомлення до системи.\n"
                    + "Аргументи:ІНДЕКС_ОРИГІНАЛУ,[НАПРЯМОК1,НАПРЯМОК2],МОВА,{Заголовок},[ТЕГ1,ТЕГ2],{СИСТЕМНІ_ОЗНАКИ}\n"
                    + "ТЕКСТ_ПОВІДОМЛЕННЯ\n"
                    + "END:\n"
                    + "Приклад: RIBBON_POST_MESSAGE:-1,[СИСТЕМА.ТЕСТ],UA,{Тестове повідомлення},[система,тест],{}\n"
                    + "Це тестове повідомлення.\n"
                    + "END:\n\n"
                    + "Примітка: команда END: на кінці повідомлення обов’язкова,\n"
                    + "без цієї команди сервер буде сприймати усі наступні команди\n"
                    + "як текст повідомлення!",
                ":-1,[СИСТЕМА.Тест],UA,{Тестове повідомлення},[тест,система],{}\nЦе тестове повідомлення на напрямок СИСТЕМА.Тест.\nEND:", 1));
        
        commands.add(new ProtocolCommand("RIBBON_POST_MESSAGE_BY_PSEUDO",
                "[LEVEL_1] RIBBON_POST_MESSAGE_BY_PSEUDO\n"
                    + "Команда випуску повідомлення до системи за псевдонапрямком.\n"
                    + "Аргументи:{ПСЕВДОНАПРЯМОК},МОВА,{Заголовок},[ТЕГ1,ТЕГ2],{СИСТЕМНІ_ОЗНАКИ}\n"
                    + "ТЕКСТ_ПОВІДОМЛЕННЯ\n"
                    + "END:\n"
                    + "Приклад: RIBBON_POST_MESSAGE_BY_PSEUDO:{Тест},UA,{Тестове повідомлення},[система,тест],{}\n"
                    + "Це тестове повідомлення.\n"
                    + "END:\n\n"
                    + "Примітка: команда END: на кінці повідомлення обов’язкова,\n"
                    + "без цієї команди сервер буде сприймати усі наступні команди\n"
                    + "як текст повідомлення!"
                    + "\n\nУВАГА! Ця команда для виконання вимагає увікнення видаленого режиму\n"
                    + "(команда RIBBON_NCTL_SET_REMOTE_MODE)!",
                ":{Тест},UA,{Тестове повідомлення},[тест,система],{}\nЦе тестове повідомлення на псевдонапрямок Тест.\nEND:", 1));
        
        commands.add(new ProtocolCommand("RIBBON_GET_MESSAGE",
                "[LEVEL_1] RIBBON_GET_MESSAGE\n"
                    + "Команда одержання тексту повідомлення певного напрямку та індексу.\n"
                    + "Аргументи:НАПРЯМОК,ІНДЕКС\n"
                    + "Приклад: RIBBON_GET_MESSAGE:СИСТЕМА.ТЕСТ,0000000107",
                ":", 1));
        
        commands.add(new ProtocolCommand("RIBBON_MODIFY_MESSAGE",
                "[LEVEL_1] RIBBON_MODIFY_MESSAGE\n"
                    + "Команда редагування тексту повідомлення певного напрямку за індексом.\n"
                    + "Аргументи: ІНДЕКС,[НАПРЯМОК1,НАПРЯМОК2],МОВА,{Заголовок},[ТЕГ1,ТЕГ2],{СИСТЕМНІ_ОЗНАКИ}\n"
                    + "НОВИЙ_ТЕКСТ_ПОВІДОМЛЕННЯ\n"
                    + "END:\n"
                    + "Приклад:RIBBON_MODIFY_MESSAGE:0000000107,[СИСТЕМА.ТЕСТ],{Тестове повідомлення},[система,тест],{}\n"
                    + "Новий текст тестового повідомлення.\n"
                    + "END:\n\n"
                    + "Примітка: команда END: на кінці повідомлення обов’язкова,\n"
                    + "без цієї команди сервер буде сприймати усі наступні команди\n"
                    + "як текст повідомлення!",
                ":-1,[],UKN,{},[],{}\n\nEND:", 1));
        
        commands.add(new ProtocolCommand("RIBBON_MODIFY_MESSAGE_BY_PSEUDO",
                "[LEVEL_1] RIBBON_MODIFY_MESSAGE_BY_PSEUDO\n"
                    + "Команда редагування тексту повідомлення певного напрямку за індексом і псевдонапрямком.\n"
                    + "Аргументи: ІНДЕКС,{ПСЕВДОНАПРЯМОК},МОВА,{Заголовок},[ТЕГ1,ТЕГ2],{СИСТЕМНІ_ОЗНАКИ}\n"
                    + "НОВИЙ_ТЕКСТ_ПОВІДОМЛЕННЯ\n"
                    + "END:\n"
                    + "Приклад:RIBBON_MODIFY_MESSAGE_BY_PSEUDO:0000000107,{Тест},{Тестове повідомлення},[система,тест],{}\n"
                    + "Новий текст тестового повідомлення.\n"
                    + "END:\n\n"
                    + "Примітка: команда END: на кінці повідомлення обов’язкова,\n"
                    + "без цієї команди сервер буде сприймати усі наступні команди\n"
                    + "як текст повідомлення!"
                    + "\n\nУВАГА! Ця команда для виконання вимагає увікнення видаленого режиму\n"
                    + "(команда RIBBON_NCTL_SET_REMOTE_MODE)!",
                ":-1,{},UKN,{},[],{}\n\nEND:", 1));
        
        commands.add(new ProtocolCommand("RIBBON_DELETE_MESSAGE",
                "[LEVEL_1] RIBBON_DELETE_MESSAGE\n"
                    + "Команда видаленя повідомлення з усіх напрямків за індексом.\n"
                    + "Аргументи:ІНДЕКС_ПОВІДОМЛЕННЯ\n"
                    + "Приклад:RIBBON_DELETE_MESSAGE:0000000107",
                ":", 1
                ));
        commands.add(new ProtocolCommand("RIBBON_ADD_MESSAGE_PROPERTY",
                "[LEVEL_1] RIBBON_ADD_MESSAGE_PROPERTY\n"
                    + "Команда додавання системної ознаки до повідомлення.\n"
                    + "Аргументи:ІНДЕКС,ТИП_ОЗНАКИ,{ТЕКСТОВА_МІТКА}\n"
                    + "Приклад:RIBBON_ADD_MESSAGE_PROPERTY:0000000107,URGENT,{!!}",
                ":-1,URGENT,{}", 1
                ));
        
        commands.add(new ProtocolCommand("RIBBON_DEL_MESSAGE_PROPERTY",
                "[LEVEL_1] RIBBON_DEL_MESSAGE_PROPERTY\n"
                    + "Команда видалення системної ознаки з повідомлення.\n"
                    + "Аргументи:ІНДЕКС,ТИП_ОЗНАКИ,ДАТА_ТА_ЧАС_ОЗНАКИ\n"
                    + "Приклад:RIBBON_DEL_MESSAGE_PROPERTY:0000000107,URGENT,23:43:27 20.01.2013",
                ":-1,URGENT,", 1
                ));
    }
    
    /**
     * Get list of commands names
     * @return string array with names of RibbonProtocol commands
     */
    public static String[] getCommandList(Integer givenLevel) {
        java.util.ListIterator<ProtocolCommand> commIter = commands.listIterator();
        java.util.ArrayList<String> commList = new java.util.ArrayList<>();
        while (commIter.hasNext()) {
            ProtocolCommand currCommand = commIter.next();
            if (currCommand.COMMAND_LEVEL == givenLevel || givenLevel == -1) {
                commList.add(currCommand.NAME);
            }
        }
        return commList.toArray(new String[commList.size()]);
    }
    
    /**
     * Get doc string of command
     * @param givenName name to search
     * @return doc of command
     */
    public static String getDoc(String givenName) {
        java.util.ListIterator<ProtocolCommand> commIter = commands.listIterator();
        while (commIter.hasNext()) {
            ProtocolCommand currCommand = commIter.next();
            if (currCommand.NAME.equals(givenName)) {
                return currCommand.DOC;
            }
        }
        return "";
    }
    
    /**
     * Get paste string of command
     * @param givenName name to search
     * @return paste string for command arguments
     */
    public static String getPaste(String givenName) {
        java.util.ListIterator<ProtocolCommand> commIter = commands.listIterator();
        while (commIter.hasNext()) {
            ProtocolCommand currCommand = commIter.next();
            if (currCommand.NAME.equals(givenName)) {
                return currCommand.AUTO_PASTE;
            }
        }
        return "";
    }
}
